### ISSUE TITLE
Proposal for `"relationships"` member at resource object top-level

### ISSUE THREAD

**Issue by ethanresnick** [json-api/json-api#618]
Here's what I'm proposing. Instead of resource objects looking like:

```
{
  attributes: {...}
  links: {
   self: "",
   tags: {...}
  }
}
```

We'd have:

```
{
  attributes: {...},
  links: {
   self: ""
  },
  relationships: {
   tags: {...}
  }
}
```

That is, the self link would go in a separate `"links"` object, and everything currently in the resource-level `"links"` would go in a new container (re)named `"relationships"`. (`"tags"` is just an example relationship in the code sample above.)

There were two things motivating this proposal:
1. Extensibility of resource-level `"links"`. It's quite likely that we may want to add other `"links"` in the future, and we can't do this post 1.0 if `"links"` and relationships share a namespace. Even pre 1.0, we can't do this without putting further restrictions on the name of a user's relationships, which I don't think we want to do. As examples of keys we might want to add to resource-level `"links"`, consider: `"collection"` (the url for the collection the resource is from), `"latest-version"`, `"alternate"`, etc. all of which are [standard](www.iana.org/assignments/link-relations.html) link-relation names that could be quite useful. I'm also thinking about how we set ourselves up to be able to implement web linking later if a common json format for it ever arises, and having a separate property bag where every official relation name is a valid key, and where the value is always a url, seems like a prerequisite.
2. Clarity, in a few different senses. First, the spec talks about "relationships" and "related resources", so having this terminology reflected in the json (much like we did for attributes), would make the spec easier to follow. Second, the `"self"` link really does seem to be a different kind of thing from a relationship, in that it can't have linkage and that a request to, e.g. `/people/1/links/self` doesn't make sense, so putting it in a different container also seems appropriate. Moreover, we'd get consistency between the format of the `"links"` object at the resource level and the top-level. (Shout out to @bintoro for making most of these points about terminology in #380. The key difference between my proposal and the discussion there is that I'm proposing to have the relationships and the links into two separate objects, where as the prior discussion left them in the same object but debated that object's name.)

Here are the effects that my proposal would have on terminology:
1. The object holding the relationships be called the "relationships" object. Each "link object" it contains would become a "relationship object". The resource-level `"links"` object, which still exists, would continue to be referred to as such.
2. Totally optionally, the `"linkage"` key could be renamed the `"content"` key (as in "the content of the relationship"), now that the parent key isn't called `"links"`. If this were done, a "linkage object" could then become a "resource identifier object", which is actually a better name because we use this form of object for things like bulk deletion too, in the `"bulk"` extension.

**Comment by bintoro** [comment]
:+1:

I support any effort to introduce a dedicated container to hold relationships.

The `links` containers are just mixed bags of stuff that could just as well be named `stuff` because content-wise they don't make sense. The spec doesn't even define what a "link" is in general because it would be useless to define such a strange idea ("a JSON-LD kinda link _or_ a JSON API relationship").

Conflating various concepts within `links` might at one point have been justified by the goal of keeping the structure very lean, but with the introduction of elaborate link objects and the `attributes` container I think we're way past that.

**Comment by cziegenberg** [comment]
Like `attributes` and the proposed `extensions` member (#574), this would help to avoid naming conflicts (not only in future versions of the spec), which I believe is very important. (Perhaps I'd call is `related` instead, because of `included`.)  :+1: 

The more proposals I see, the more I think that the whole spec should be checked for possible naming conflicts in the current and in future versions of the spec, as well as other types of known conflicts, like my proposals regarding derived attributes/insufficient permissions (#575, PR #589 by @hhware). IMHO it should be the primary goal for a version 1.0, to ensure that future versions have a chance to be compatible to this version.

**Comment by hhware** [comment]
A related proposal: #622.

**Comment by dgeb** [comment]
@ethanresnick Could you please provide a more complete example? Perhaps a reworking of the home page example? I'm especially interested in understanding where you're moving the contents (`self`, `related`, etc.) of link objects.

**Comment by tkellen** [comment]
I'm wholeheartedly :+1: to this.

**Comment by ethanresnick** [comment]
@bintoro @ziege @tkellen Glad to hear you guys are on board with this! :)

@dgeb Absolutely! Here's how the home page example would look:
1. First, under the limited version of my proposal, which keeps the contents of what are currently called link objects exactly as is:
   
   ```
   {
   "links": {
     "self": "http://example.com/posts",
     "next": "http://example.com/posts?page[offset]=2",
     "last": "http://example.com/posts?page[offset]=10"
   },
   "data": [{
     "type": "posts",
     "id": "1",
     "attributes": {
       "title": "JSON API paints my bikeshed!"
     },
     "relationships": {
       "author": {
         "self": "http://example.com/posts/1/links/author",
         "related": "http://example.com/posts/1/author",
         "linkage": { "type": "people", "id": "9" }
       },
       "comments": {
         "self": "http://example.com/posts/1/links/comments",
         "related": "http://example.com/posts/1/comments",
         "linkage": [
           { "type": "comments", "id": "5" },
           { "type": "comments", "id": "12" }
         ]
       }
     },
     "links": {
       "self": "http://example.com/posts/1"
     }
   }]
   }
   ```
   
   Quite literally all I have done in this version is rename the resource-level `"links"` key to `"relationships"`, then pull the `"self"` link out of that bag and put it into a resource-level bag called `"links"`. Again, the motivation is to make that resource-level `"links"` extensible over time, without conflicting with user-defined relationship names, and perhaps to clear up the terminology as a result.
   &nbsp;
2. The version below is an exact copy of the above, except that I renamed the `"linkage"` key to `"content"`. See my original proposal (the second bullet under effects on terminology) for more on reasoning/terminology ramifications of this.
   
   ```
   {
   "links": {
     "self": "http://example.com/posts",
     "next": "http://example.com/posts?page[offset]=2",
     "last": "http://example.com/posts?page[offset]=10"
   },
   "data": [{
     "type": "posts",
     "id": "1",
     "attributes": {
       "title": "JSON API paints my bikeshed!"
     },
     "relationships": {
       "author": {
         "self": "http://example.com/posts/1/links/author",
         "related": "http://example.com/posts/1/author",
         "content": { "type": "people", "id": "9" }
       },
       "comments": {
         "self": "http://example.com/posts/1/links/comments",
         "related": "http://example.com/posts/1/comments",
         "content": [
           { "type": "comments", "id": "5" },
           { "type": "comments", "id": "12" }
         ]
       }
     },
     "links": {
       "self": "http://example.com/posts/1"
     }
   }]
   }
   ```
   
   &nbsp;
3. Last, and I didn't propose this initially because I wanted to see if people were on board with the general idea first, we could amend version (2) above. This amendment would group the keys that are in "link objects" (which, as mentioned, would be renamed "relationship objects") and that contain urls into a `"links"` bag within the relationship object. This, like the renaming of `"linkage"` to `"content"` is totally optional from the perspective of the core proposal (bullet 1), but has the advantage of unifying the presentation of keys containing urls at every level (i.e. at the top-level, within resource objects, and within "relationship objects"). That would look like this:
   
   ```
   {
   "links": {
     "self": "http://example.com/posts",
     "next": "http://example.com/posts?page[offset]=2",
     "last": "http://example.com/posts?page[offset]=10"
   },
   "data": [{
     "type": "posts",
     "id": "1",
     "attributes": {
       "title": "JSON API paints my bikeshed!"
     },
     "relationships": {
       "author": {
         "links": {
           "self": "http://example.com/posts/1/links/author",
           "related": "http://example.com/posts/1/author"
         },
         "content": { "type": "people", "id": "9" }
       },
       "comments": {
         "links": {
           "self": "http://example.com/posts/1/links/comments",
           "related": "http://example.com/posts/1/comments"
         },
         "content": [
           { "type": "comments", "id": "5" },
           { "type": "comments", "id": "12" }
         ]
       }
     },
     "links": {
       "self": "http://example.com/posts/1"
     }
   }]
   }
   ```

**Comment by hhware** [comment]
@ethanresnick, I fully support all items including the 3rd one, which is also in line with #622. I guess extraction of of attributes into a separate object (#588) set us both to think in the same direction...

**Comment by dgeb** [comment]
@ethanresnick I favor the semantics and structure of the third option because it maintains links related to resources in their own `links` object. The other options maintain blurred lines between links and relationships. As @hhware notes, it's also in line with #622, so it's good to see your thinking converging.

If we are to go down this route, we should also rethink the relationship URL recommendations for consistency. Adding `links` before the relationship name would make much less sense. I'd prefer that the structure of URLs match the document structure closely.

I'm not 100% onboard with the following, but I think it's getting there:

```
{
  "links": {
    "self": "http://example.com/posts",
    "next": "http://example.com/posts?page[offset]=2",
    "last": "http://example.com/posts?page[offset]=10"
  },
  "data": [{
    "type": "posts",
    "id": "1",
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    },
    "relationships": {
      "author": {
        "links": {
          "related": "http://example.com/posts/1/author",
          "linkage": "http://example.com/posts/1/author/linkage"
        },
        "linkage": { "type": "people", "id": "9" }
      },
      "comments": {
        "links": {
          "related": "http://example.com/posts/1/comments",
          "linkage": "http://example.com/posts/1/comments/linkage"
        },
        "linkage": [
          { "type": "comments", "id": "5" },
          { "type": "comments", "id": "12" }
        ]
      }
    },
    "links": {
      "self": "http://example.com/posts/1"
    }
  }]
}
```

This removes the `self` member from the `links` object in a relationship object in favor of the less ambiguous `linkage`. It also changes the recommendation for what I might now call a "linkage URL". I'm not positive that I like these tweaks exactly as proposed. I think this area needs refinement and would welcome further comments.

@ethanresnick since this was your proposal, it's of course up to you if you want to put the work into a PR. @tkellen and I are planning to work all day Monday on the spec. Obviously, time is short before 1.0 on the 21st. Let me know if your schedule permits working on this - otherwise Tyler and I can try to refine these concepts and test them with our implementations. Either way, thanks for your work on this. Also, thanks to @hhware for similar work in #622.

**Comment by ethanresnick** [comment]
@dgeb Thanks for the feedback! The third option is my preferred one as well, and I'll work up a PR tonight or tomorrow! 

I hadn't thought about how these changes would influence the linkage url. My inclination would be to keep the key in the `"links"` object called `"self"`, because "self" is an official and appropriate link relation, which'll be really important if we later try to define a mapping between the `"links"` object and RFC 5988. But I would change the name of the path in the url from "links" to "relationships", so a relationship object would look like this:

```
{ 
  "links": {
    "self": "http://example.com/posts/1/relationships/author",
    "related": "http://example.com/posts/1/relationships/author"
  },
  "linkage": {"type": "people", "id": "9"}
}
```

Thoughts?

**Comment by dgeb** [comment]
@ethanresnick I think you meant to keep the `related` URL as `http://example.com/posts/1/author` in your example, right?

If so, then that sounds like a great starting point for a PR. Thanks in advance!

**Comment by ethanresnick** [comment]
Yup, that's what I meant. Just, overly ambitious copy and pasting :) I'll start work on a PR soon.

Also, I know we're coming up on the finish line for JSON API 1.0 and there's still a lot of final touches to get done. So I just wanted to offer to help out in any way that would be useful. I'll be available much of Monday, so could help with writing up last-minute changes, closing resolved issues, or discussing last-minute design decisions. Again, just let me know whatever would be most helpful, and feel free to get in touch by email: ethan.resnick@gmail.com As you've no doubt surmised, I've become pretty invested in the spec and want to see it be as good as it can be!

**Comment by hhware** [comment]
@dgeb, @ethanresnick, I also support the choice of `self` as key, and replacement of `links` with `relationships` in the URL seems to be right to the point.

**Comment by dgeb** [comment]
Implemented and merged in #625.

**Comment by suportjinhub** [comment]
> Here's what I'm proposing. Instead of resource objects looking like:
> 
> ```
> {
>   attributes: {...}
>   links: {
>    self: "",
>    tags: {...}
>   }
> }
> ```
> 
> We'd have:
> 
> ```
> {
>   attributes: {...},
>   links: {
>    self: ""
>   },
>   relationships: {
>    tags: {...}
>   }
> }
> ```
> 
> That is, the self link would go in a separate `"links"` object, and everything currently in the resource-level `"links"` would go in a new container (re)named `"relationships"`. (`"tags"` is just an example relationship in the code sample above.)
> 
> There were two things motivating this proposal:
> 
> 1. Extensibility of resource-level `"links"`. It's quite likely that we may want to add other `"links"` in the future, and we can't do this post 1.0 if `"links"` and relationships share a namespace. Even pre 1.0, we can't do this without putting further restrictions on the name of a user's relationships, which I don't think we want to do. As examples of keys we might want to add to resource-level `"links"`, consider: `"collection"` (the url for the collection the resource is from), `"latest-version"`, `"alternate"`, etc. all of which are [standard](www.iana.org/assignments/link-relations.html) link-relation names that could be quite useful. I'm also thinking about how we set ourselves up to be able to implement web linking later if a common json format for it ever arises, and having a separate property bag where every official relation name is a valid key, and where the value is always a url, seems like a prerequisite.
> 2. Clarity, in a few different senses. First, the spec talks about "relationships" and "related resources", so having this terminology reflected in the json (much like we did for attributes), would make the spec easier to follow. Second, the `"self"` link really does seem to be a different kind of thing from a relationship, in that it can't have linkage and that a request to, e.g. `/people/1/links/self` doesn't make sense, so putting it in a different container also seems appropriate. Moreover, we'd get consistency between the format of the `"links"` object at the resource level and the top-level. (Shout out to [@bintoro](https://github.com/bintoro) for making most of these points about terminology in [Rename top-level 'links' & clarify what a link is #380](https://github.com/json-api/json-api/issues/380). The key difference between my proposal and the discussion there is that I'm proposing to have the relationships and the links into two separate objects, where as the prior discussion left them in the same object but debated that object's name.)
> 
> Here are the effects that my proposal would have on terminology:
> 
> 1. The object holding the relationships be called the "relationships" object. Each "link object" it contains would become a "relationship object". The resource-level `"links"` object, which still exists, would continue to be referred to as such.
> 2. Totally optionally, the `"linkage"` key could be renamed the `"content"` key (as in "the content of the relationship"), now that the parent key isn't called `"links"`. If this were done, a "linkage object" could then become a "resource identifier object", which is actually a better name because we use this form of object for things like bulk deletion too, in the `"bulk"` extension.

{
  attributes: {...}
  links: {
   self: "",
   tags: {...}
  }
}
